codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -244,20 +246,41 @@ class KotlinJmespathExpressionVisitor(
     }
 
     override fun visitMultiSelectHash(expression: MultiSelectHashExpression): VisitedExpression {
-        throw CodegenException("MultiSelectHashExpression is unsupported")
+        val properties = mutableListOf<String>()
+        expression.expressions.forEach {
+            properties.add("val ${it.key}: T")
+        }
+        writer.write("class Selection<T>(${properties.joinToString()})")
+
+        val listName = bestTempVarName("multiSelect")
+        writer.openBlock("val #L = listOfNotNull(", listName)
+        writer.openBlock("run {")
+
+        val identifiers = mutableListOf<String>()
+        expression.expressions.forEach {
+            val identifier = addTempVar(it.key, it.value.accept(this).identifier)
+            identifiers.add(identifier)
+        }
+        writer.write("Selection(${identifiers.joinToString()})")
```

Comment:
**Nit**: Also could be simplified to avoid mutability:

```kotlin
val identifiers = expression.expressions.joinToString { addTempVar(it.key, it.value.accept(this).identifier) }
writer.write("Selection($identifiers)")
```